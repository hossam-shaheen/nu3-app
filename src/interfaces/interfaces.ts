export interface ProductType {
    id: number;
    title: string;
    price: number;
    image: string;
    vendor: string;
    link: string;
}

export interface ProductsType {
    totalItems: number;
    items: ProductType[];
}

export interface ProductsSearchProps {
    onSearch: (searchKeyWord: string) => void;
}

export interface SortProps {
    onSort: (sortKeyWord: string) => void;
}
export interface ImageType {
    id: number;
    src: string;
    alt: string;
}
export interface ProductAttributeProps {
    productImage: ImageType;
    productImages: ImageType[];
    ProductTitle: string;
    productBodyHtml: string;
    productVendor: string;
}


export interface ProductVariantsProps {
    options: OptionsType[];
    variants: VariantType[];
}


export interface ProductInfoType {
    id: number;
    title: string;
    body_html: string;
    vendor: string;
    images: ImageType[];
    image: ImageType;
    options: OptionsType[];
    variants: VariantType[];
    url: string;
}

export interface VariantType {
    id: number,
    title: string,
    price: number,
}

export interface OptionsType {
    id: number,
    values: string[],
    name: string,
}

export interface ErrorType {
    icon?: string,
    message: string
}