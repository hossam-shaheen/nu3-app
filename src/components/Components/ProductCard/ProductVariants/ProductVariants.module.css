.product-variants {
    width: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    margin-top: 20px;
    margin-bottom: 20px;
}

.product-variants select {
    width: calc(100% - 155px);
    height: 45px;
}

.product-price {
    font-size: 25px;
    font-weight: 600;
    color: var(--secondary);
    margin: 0;
}