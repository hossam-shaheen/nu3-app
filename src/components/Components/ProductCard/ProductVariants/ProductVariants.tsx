import { FunctionComponent, useState, ChangeEvent, useEffect } from 'react';
import { OptionsType, ProductVariantsProps, VariantType } from 'src/interfaces/interfaces';
import classes from './ProductVariants.module.css';


export const ProductVariants: FunctionComponent<{ productVariants: (ProductVariantsProps | null) }> = ({ productVariants }): JSX.Element => {
    const [selectedOption, setSelectedOption] = useState<string | undefined>();
    const [selectedVariant, setSelectedVariant] = useState<VariantType | undefined>();

    const optionValues = productVariants?.options.map((option: OptionsType) => option.values).flat(1) || [];

    useEffect((): void => {
        optionValues && setSelectedOption(optionValues[0]);
        //eslint-disable-next-line
    }, []);

    useEffect((): void => {
        setSelectedVariant(productVariants?.variants.find((variant: VariantType) => variant.title === selectedOption));
        //eslint-disable-next-line
    }, [selectedOption]);

    const onSelectVariants = (e: ChangeEvent<HTMLSelectElement>) => {
        setSelectedOption(e.target.value);
    }

    return <div className={classes["product-variants"]}>
        {optionValues.length > 0 && <select value={selectedOption} onChange={onSelectVariants} data-testid="product-variants-select">
            {optionValues?.length > 0 && optionValues.map((option) => {
                return <option key={option} value={option}>
                    {option}
                </option>
            })}
        </select>}
        <div className={classes["product-price"]}>
            <p>&#x20AC; {selectedVariant?.price}</p>
        </div>
    </div>
}
export default ProductVariants;