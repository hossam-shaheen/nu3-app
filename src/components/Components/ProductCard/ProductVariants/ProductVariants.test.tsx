import { render, screen, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { ProductVariantsProps } from "src/interfaces/interfaces";
import ProductVariants from "./ProductVariants";

const mockedProductVariants: ProductVariantsProps = {
  variants: [
    {
      id: 1,
      title: "150 g",
      price: 18.99,
    },
    {
      id: 2,
      title: "3 x 150 g",
      price: 51.99,
    },
  ],
  options: [
    {
      id: 1,
      name: "amount",
      values: ["150 g", "3 x 150 g"],
    },
  ],
};
const validateProductVariantsAndPrice = (selection: string, price: number) => {
  const productSelectVariants = screen.getByTestId("product-variants-select");

  expect(productSelectVariants).toHaveValue(selection);
  const productPrice = screen.getByText(`${price}`, { exact: false });
  expect(productPrice).toBeInTheDocument();
};

test("ProductVariants pass productVariants Props : Renders", () => {
  render(<ProductVariants productVariants={mockedProductVariants} />);

  const productSelectVariants = screen.getByTestId("product-variants-select");
  expect(productSelectVariants).toBeInTheDocument();

  validateProductVariantsAndPrice("150 g", 18.99);
});

test(`ProductVariants didn't pass productVariants Props : Renders`, () => {
  render(<ProductVariants productVariants={null} />);

  const productSelectVariants = screen.queryByTestId("product-variants-select");
  expect(productSelectVariants).not.toBeInTheDocument();
});

test("ProductVariants change select Variants  : Renders", () => {
  render(<ProductVariants productVariants={mockedProductVariants} />);

  const productSelectVariants = screen.getByTestId("product-variants-select");

  validateProductVariantsAndPrice("150 g", 18.99);

  fireEvent.change(productSelectVariants, { target: { value: "3 x 150 g" } });

  validateProductVariantsAndPrice("3 x 150 g", 51.99);

  fireEvent.change(productSelectVariants, { target: { value: "150 g" } });

  validateProductVariantsAndPrice("150 g", 18.99);
});

test(`ProductVariants pass ProductAttributes Props: Snapshot`, () => {
  const component = renderer.create(
    <ProductVariants productVariants={mockedProductVariants} />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test(`ProductVariants didn't pass productVariants Props : Snapshot`, () => {
  const component = renderer.create(<ProductVariants productVariants={null} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
