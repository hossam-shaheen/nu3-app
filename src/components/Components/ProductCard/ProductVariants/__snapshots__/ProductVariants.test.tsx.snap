// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductVariants didn't pass productVariants Props : Snapshot 1`] = `
<div
  className="product-variants"
>
  <div
    className="product-price"
  >
    <p>
      € 
    </p>
  </div>
</div>
`;

exports[`ProductVariants pass ProductAttributes Props: Snapshot 1`] = `
<div
  className="product-variants"
>
  <select
    data-testid="product-variants-select"
    onChange={[Function]}
  >
    <option
      value="150 g"
    >
      150 g
    </option>
    <option
      value="3 x 150 g"
    >
      3 x 150 g
    </option>
  </select>
  <div
    className="product-price"
  >
    <p>
      € 
    </p>
  </div>
</div>
`;
